models/user.js

```diff
@@ -25,6 +26,55 @@ async function findAll() {
   return results.rows;
 }
 
+async function findAllWithPagination(values) {
+  const offset = (values.page - 1) * values.per_page;
+
+  const query = {
+    text: `
+      WITH user_window AS (
+        SELECT
+          COUNT(*) OVER()::INTEGER as total_rows,
+          id
+        FROM users
+        ORDER BY updated_at DESC
+        LIMIT $1 OFFSET $2
+      )
+
+      SELECT
+        *
+      FROM
+        users
+      INNER JOIN
+        user_window ON users.id = user_window.id
+      CROSS JOIN LATERAL (
+        SELECT
+          get_user_current_tabcoins(users.id) as tabcoins,
+          get_user_current_tabcash(users.id) as tabcash
+      ) as balance
+      ORDER BY updated_at DESC
+    `,
+    values: [values.limit || values.per_page, offset],
+  };
+
+  const queryResults = await database.query(query);
+
+  const results = {
+    rows: queryResults.rows,
+  };
+
+  values.totalRows = results.rows[0]?.total_rows ?? (await countTotalRows());
+
+  results.pagination = await pagination.getPagination(values);
+
+  return results;
+}
+
+async function countTotalRows() {
```

Comment:
Em qual situação a `countTotalRows()` é necessária?

Reply:
Quando a consulta não retorna resultados. Um exemplo é no teste `With a "page" out of bounds`.